paymaster/build_txn.go

```diff
@@ -0,0 +1,350 @@
+type BuildTransactionRequest struct {
+	// The transaction to be executed by the paymaster
+	Transaction *UserTransaction `json:"transaction"`
+	// Execution parameters to be used when executing the transaction
+	Parameters *UserParameters `json:"parameters"`
+}
+
+// UserTransaction represents a user transaction (deploy, invoke,
+// or deploy_and_invoke).
+type UserTransaction struct {
+	// The type of the transaction to be executed by the paymaster
+	Type UserTxnType `json:"type"`
+	// The deployment data for the transaction, used for `deploy` and
+	// `deploy_and_invoke` transaction types.
+	// Should be `nil` for `invoke` transaction types.
+	Deployment *AccDeploymentData `json:"deployment,omitempty"`
+	// The invoke data for the transaction, used for `invoke` and
+	// `deploy_and_invoke` transaction types.
+	// Should be `nil` for `deploy` transaction types.
+	Invoke *UserInvoke `json:"invoke,omitempty"`
+}
+
+// An enum representing the type of the transaction to be executed
+// by the paymaster
+type UserTxnType string
+
+const (
+	// Represents a deploy transaction
+	UserTxnDeploy UserTxnType = "deploy"
+	// Represents an invoke transaction
+	UserTxnInvoke UserTxnType = "invoke"
+	// Represents a deploy and invoke transaction
+	UserTxnDeployAndInvoke UserTxnType = "deploy_and_invoke"
+)
+
+// MarshalJSON marshals the UserTxnType to JSON.
+func (u UserTxnType) MarshalJSON() ([]byte, error) {
+	switch u {
+	case UserTxnDeploy, UserTxnInvoke, UserTxnDeployAndInvoke:
+		return json.Marshal(string(u))
+	}
+
+	return nil, fmt.Errorf("invalid user transaction type: %s", u)
+}
+
+// UnmarshalJSON unmarshals the JSON data into a UserTxnType.
+func (u *UserTxnType) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	switch s {
+	case "deploy":
+		*u = UserTxnDeploy
+	case "invoke":
+		*u = UserTxnInvoke
+	case "deploy_and_invoke":
+		*u = UserTxnDeployAndInvoke
+	default:
+		return fmt.Errorf("invalid user transaction type: %s", s)
+	}
+
+	return nil
+}
+
+// Data required to deploy an account at an address.
+type AccDeploymentData struct {
+	// The expected address to be deployed, used to double check
+	Address *felt.Felt `json:"address"`
+	// The hash of the deployed contract's class
+	ClassHash *felt.Felt `json:"class_hash"`
+	// The salt used for the contract address calculation
+	Salt *felt.Felt `json:"salt"`
+	// The parameters passed to the constructor
+	ConstructorCalldata []*felt.Felt `json:"calldata"`
+	// Optional array of felts to be added to the signature
+	SignatureData []*felt.Felt `json:"sigdata,omitempty"`
+	// The Cairo version of the account contract (e.g. 1 or 2). Cairo 0 is not supported.
+	Version uint8 `json:"version"`
+}
+
+// Calls to be executed by the paymaster and the user account address that will be called
+type UserInvoke struct {
+	// The address of the user account
+	UserAddress *felt.Felt `json:"user_address"`
```

Comment:
Similar comment here, it shouldn't be called `UserAddress *felt.Felt`, as far as I can see the type definition should be
```golang
type UserInvoke struct { 
  User felt.Address `json:...`
  ...
}
```